### DISCUSSION TITLE
Multiple (rapid) requests for refresh token (following documentation)

### DISCUSSION THREAD

**Discussion by avisra** [nextauthjs/next-auth#1455] [Help]
**Describe the bug**
I setup the refresh token rotation copying the code form the tutorials section. However, on page refresh after the expiration... my refresh token is failing, and by adding console.logs to the different areas in the code, I determined that it is calling refreshAccessTokens two times (back to back, rapid fire, with the same refresh token). The first response is successful - the second response is not. And its the second response which fails everything because after the first request - it invalidates the refresh token - so the second one fails.

**Steps to reproduce**
Setup the refresh token example with IdentityServer4 as the provider.

**Additional context**
Using IdentityServer4 as the provider

![image](https://user-images.githubusercontent.com/433538/110198526-2bdf3100-7e21-11eb-910f-ae3f44fce35f.png)

**Comment by balazsorban44** [comment]
would you be able to share some of your code/configuration?

We are using this technique at work together with IDS4 and works fine. So it must be some misconfiguration on your side.

400 means bad request, so might be sending something that is not supposed to be sent or something similar.

**Comment by avisra** [comment]
The 400 error is because the second request is attempting to use an already redeemed refresh token. This behaviour is experienced when configuring the client in Identity Server to only allow one-time refresh tokens.

RefreshTokenUsage = OneTimeOnly
http://docs.identityserver.io/en/latest/topics/refresh_tokens.html

I believe the issue is the double call to get the refresh token (maybe a double call to JWT callback in NextAuth?) I will put together a project demoing the issue and follow up on the thread here

**Reply by balazsorban44** [reply]
We are also using `RefreshTokenUsage = OneTimeOnly`, so still not convinced that it isn't just a misconfiguration.

**Reply by avisra** [reply]
Here is our configuration. I hope this is enough - otherwise I will have to setup a repo in the next day or two. I've lowered the value of the access token expiration to 60 second for testing. The issue does **not** reproduce when the keepalive triggers the JWT callback - the refresh token works successfully. I can only reproduce the issue when **doing a page refresh**. After refreshing, my refresh token call happens twice, in very quick succession - quick enough to where the second one starts before the first one is able to update the refresh token. It appears that if I page refresh after the token has expired - but before the keepalive triggers the JWT - JWT callback is double triggered. It appears to also happen if I leave the site and come back later (after the access token is expired, but refresh token has not). I hope this is enough information

And of course, I hope this is a simple configuration issue and not a bug.

**App.js configuration (CommerceProvider is part of BigCommerce data hooks)**
```jsx
    <Provider
      // Provider options are not required but can be useful in situations where
      // you have a short session maxAge time. Shown here with default values.
      options={{
        // Client Max Age controls how often the useSession in the client should
        // contact the server to sync the session state. Value in seconds.
        // e.g.
        // * 0  - Disabled (always use cache value)
        // * 60 - Sync session state with server if it's older than 60 seconds
        clientMaxAge: 0,
        // Keep Alive tells windows / tabs that are signed in to keep sending
        // a keep alive request (which extends the current session expiry) to
        // prevent sessions in open windows from expiring. Value in seconds.
        //
        // Note: If a session has expired when keep alive is triggered, all open
        // windows / tabs will be updated to reflect the user is signed out.
        keepAlive: 10, // value for testing
      }}
      session={pageProps.session}
    >
      <CommerceProvider locale={locale}>
        <Component
          {...pageProps}
          profile={profile}
          setProfile={setProfile}
          isDesigner={isDesigner}
        />
      </CommerceProvider>
    </Provider>
```

**NextAuth API:**
```js
import NextAuth from "next-auth";
import Providers from "next-auth/providers";
import { getAccountStatus, processShopLogin } from "../../../services/login";
import axios from "axios";

export default (req, res) => {
  async function refreshAccessToken(req, res, token) {
    try {
      if ((await getAccountStatus(token.accessToken)) !== -1) {
        // Use account exists, continue with token refresh
        // Here we avoid refreshing tokens for sessions of users which have been deleted from our system
        const url = `${process.env.IDENTITY_AWS_URL}/connect/token`;
        const params = new URLSearchParams();
  
        params.append("client_id", "nextjs");
        params.append("client_secret", "secret");
        params.append("grant_type", "refresh_token");
        params.append("refresh_token", token.refreshToken);
  
        console.log("posting to ids for refresh token");
  
        const refreshResponse = await axios.post(url, params, {
          headers: {
            "Content-Type": "application/x-www-form-urlencoded",
          },
        });
  
        const refreshedTokens = refreshResponse?.data;
  
        if (refreshResponse.status !== 200) {
          throw refreshedTokens;
        }
  
        // .then(async (res) => {
        //   console.log("refresh", res);
        //   token.expires = new Date(
        //     Date.now() + res.data.expires_in * 1000 - 10
        //   );
        //   token.accessToken = res.data.access_token;
        //   token.refreshToken = res.data.refresh_token;
        // });
  
        // Renew shop cookies
        await processShopLogin(req, res, token);
        console.log("refresh done!!", refreshedTokens);
        return {
          ...token,
          accessToken: refreshedTokens.access_token,
          accessTokenExpires: Date.now() + refreshedTokens.expires_in * 1000,
          refreshToken: refreshedTokens.refresh_token ?? token.refreshToken, // Fall back to old refresh token
        };
      }
    } catch (error) {
      console.log(error);
  
      return {
        ...token,
        error: "RefreshAccessTokenError",
      };
    }
  }

  return NextAuth(
    req,
    res,
    // For more information on each option (and a full list of options) go to
    // https://next-auth.js.org/configuration/options
    {
      // https://next-auth.js.org/configuration/providers
      providers: [
        // Note - Identity Server has a 5 minute clock skew - so if a token is supposed to expire in 10 minutes, it won't REALLY expire for 15 minutes.
        Providers.IdentityServer4({
          id: "nextjs",
          name: "nextjs",
          scope: "openid profile email apis ids:profile offline_access", // Allowed Scopes
          domain: process.env.IDENTITY_HOST,
          clientId: "nextjs",
          protection: "pkce",
          clientSecret: process.env.IDENTITY_CLIENT_SECRET,
          authorizationUrl: `https://${process.env.IDENTITY_HOST}/connect/authorize?response_type=code&prompt=login`,
        }),
      ],
      // Database optional. MySQL, Maria DB, Postgres and MongoDB are supported.
      // https://next-auth.js.org/configuration/database
      //
      // Notes:
      // * You must to install an appropriate node_module for your database
      // * The Email provider requires a database (OAuth providers do not)
      //database: process.env.DATABASE_URL,

      // The secret should be set to a reasonably long random string.
      // It is used to sign cookies and to sign and encrypt JSON Web Tokens, unless
      // a seperate secret is defined explicitly for encrypting the JWT.
      secret: "INp8IvdIyeMcoGAgFGoA61DdBglwwSqnXJZkgz8PSnw", //process.env.SECRET,

      session: {
        // Use JSON Web Tokens for session instead of database sessions.
        // This option can be used with or without a database for users/accounts.
        // Note: `jwt` is automatically set to `true` if no database is specified.
        jwt: true,

        // Seconds - How long until an idle session expires and is no longer valid.
        //maxAge: 3600, // 1 hour to match identity server timer

        // Seconds - Throttle how frequently to write to database to extend a session.
        // Use it to limit write operations. Set to 0 to always update the database.
        // Note: This option is ignored if using JSON Web Tokens
        // updateAge: 24 * 60 * 60, // 24 hours
      },

      // JSON Web tokens are only used for sessions if the `jwt: true` session
      // option is set - or by default if no database is specified.
      // https://next-auth.js.org/configuration/options#jwt
      jwt: {
        // A secret to use for key generation (you should set this explicitly)
        secret: "INp8IvdIyeMcoGAgFGoA61DdBglwwSqnXJZkgz8PSnw",

        // Set to true to use encryption (default: false)
        // encryption: true,

        // You can define your own encode/decode functions for signing and encryption
        // if you want to override the default behaviour.
        // encode: async ({ secret, token, maxAge }) => {},
        // decode: async ({ secret, token, maxAge }) => {},
      },

      // You can define custom pages to override the built-in pages.
      // The routes shown here are the default URLs that will be used when a custom
      // pages is not specified for that route.
      // https://next-auth.js.org/configuration/pages
      pages: {
        // signIn: '/api/auth/signin',  // Displays signin buttons
        // signOut: '/api/auth/signout', // Displays form with sign out button
        // error: '/api/auth/error', // Error code passed in query string as ?error=
        // verifyRequest: '/api/auth/verify-request', // Used for check email page
        // newUser: null // If set, new users will be directed here on first sign in
      },

      // Callbacks are asynchronous functions you can use to control what happens
      // when an action is performed.
      // https://next-auth.js.org/configuration/callbacks
      callbacks: {
        signIn: async (user, account, profile) => {
          user.image = profile?.picture;
          // user.first_name = profile.given_name;
          // user.last_name = user.family_name;
          return Promise.resolve(true);
        },
        redirect: async (url, baseUrl) => {
          if (
            url.startsWith(`${process.env.SITE_URL}/api/auth/signin-callback`)
          ) {
            return Promise.resolve(url);
          } else {
            const callbackUrl = url.startsWith(baseUrl) ? url : baseUrl;

            return Promise.resolve(
              `${process.env.SITE_URL}/api/auth/signin-callback${
                callbackUrl && callbackUrl.length
                  ? `?callbackUrl=${encodeURIComponent(callbackUrl)}`
                  : ""
              }`
            );
          }
        },
        async session(session, token) {
          if(token){
            session.user = token.user;
            session.accessToken = token.accessToken;
            session.error = token.error;
            session.isProfileComplete = token.isProfileComplete;
          }

          return session;
        },
        async jwt (token, user, account, profile, isNewUser) {
          //console.log("server response", res);
          // If user exists, this means that the user is signing in
          if (account && user) {
            // If the user is this far, we can assume they exist in Identity Server already.
            let isProfileComplete = false;
            const result = await getAccountStatus(account.accessToken);
            console.log(result);
            if (result === -1) {
              console.log(
                "something went wrong in getAccountStatus, failing the login"
              );
              // Something went wrong in our getAccountStatus method, fail the login
              return Promise.reject();
            } else if (result === 1) {
              console.log("the account existed and has completed its profile");
              isProfileComplete = true;
            } else if (result === 0) {
              console.log(
                "the account existed, but their profile has not been completed"
              );
              isProfileComplete = false;
            }

            // continue with processShopLogin
            await processShopLogin(req, res, account);
            console.log(account);
            return {
              accessToken: account.accessToken,
              accessTokenExpires: Date.now() + account.expires_in * 1000,
              refreshToken: account.refresh_token,
              isProfileComplete: isProfileComplete,
              user,
            };
          }

          // Return previous token if the access token has not expired yet
          if ((Date.now() < token.accessTokenExpires)) {
            return token;
          }

          return await refreshAccessToken(req, res, token);
        },
      },
    }
  );
}
```

**IdentityServer 4 client configuration**
```
new Client
{
    ClientName = "NEXTJS App",
    ClientId = "nextjs",
    AllowedGrantTypes = GrantTypes.Code,
    RequirePkce = true,
    RequireClientSecret = true,
    ClientSecrets = { new Secret("secret".Sha256()) }, // TEMPORARY
    AllowedCorsOrigins = { "http://localhost:3000" },
    AllowOfflineAccess = true, // Allows refresh tokens
    AccessTokenLifetime = 60,
    RefreshTokenUsage = TokenUsage.OneTimeOnly,
    AllowedScopes = {
        IdentityServerConstants.StandardScopes.OpenId,
        IdentityServerConstants.StandardScopes.Profile,
        IdentityServerConstants.StandardScopes.Email,
        "apis",
        "ids:profile"
    },
    RedirectUris = {
        "http://localhost:3000/api/auth/callback/nextjs"
    },
    PostLogoutRedirectUris = { "http://localhost:3000/" },
},
```

**Reply by balazsorban44** [reply]
Okay, so:
> It appears to also happen if I leave the site and come back later (after the access token is expired, but the refresh token has not).

Made me curious, there is a chance that we also experience this at work. I remember someone reporting issues with token rotation while leaving the computer idle and coming back...
An isolated reproduction here would help tremendously here!

Since the refresh token implementation is in user-land right now, there might be a bug there. Hopefully, once we get the built-in token rotation, this won't be a problem. Unfortunately, it is not something done quickly, so let's, for now, fix it in the user-land implementation. I assume you followed this guide: https://next-auth.js.org/tutorials/refresh-token-rotation

I'll carefully double-check the flow, until then, a complete repro repository would be awesome!


The problem might originate from the following wrong assumption:

```js
 // Return previous token if the access token has not expired yet
      if (Date.now() < token.accessTokenExpires) {
        return token;
      }

      // Access token has expired, try to update it
      return refreshAccessToken(token);
```

As you mentioned, the refresh token has its own expiry, so it cannot use the access token expiry(?)

**Reply by balazsorban44** [reply]
So this started bugging me, as I think I could resolve our problem at work if we figured out the issue here... @avisra feel free to reach out on Twitter or e-mail, I would like to have a better conversation about this!

**Reply by avisra** [reply]
@balazsorban44 I am struggling to find the time to get the repro setup, but it is still on my radar! I'll try to get this done soon.

**Reply by balazsorban44** [reply]
That would be awesome, thank you!

**Reply by maxbec** [reply]
Hello guys... I think we ware facing the same issue? Was there any solution?

**Reply by JSONRice** [reply]
@balazsorban44 any ideas? Seems like this is still an issue.

**Comment by makezi** [comment]
My work seems to be facing the same issue and @balazsorban44, you seem to be onto something in regards to the `accessTokenExpires` not updating properly.

I've attempted to reproduce this ([repo here](https://github.com/makezi/next-auth-refresh-token)) which uses [Demo IdentityServer4](https://demo.identityserver.io/) as the auth provider and I've outlined the steps in the `README` as best as I can. 

If you have a look, you can see that in the first attempt to retrieve the new tokens (by clicking the `server` page after > 75 seconds and checking server logs) that a new `accessTokenExpires` value has been set (which shows up in token in the `session` callback). If we click the `server` page again, it will seem to want to do another attempt to retrieve the new tokens BUT you can see that the token from the `session` callback has the `accessTokenExpires` still set from the initial value - it didn't seem to update properly.

I hope this helps.

**Reply by balazsorban44** [reply]
Will check on this, thanks!

**Reply by makezi** [reply]
Hey @balazsorban44 had a chance to look into this? Looking into it again, seems like the jwt callback doesn't like setting new keys for the token via `refreshAccessToken` - As mentioned, in the first return from `accessTokenExpires` will have the new expire time set but any other future jwt callbacks it will just revert back to the initial value.

**Reply by balazsorban44** [reply]
not yet unfortunately, as this issue disappeared in our application. 😳

**Reply by makezi** [reply]
Interesting. Which version of 'next-auth' are you using and is your work using the refresh token in the tutorials verbatim? My work is currently on 3.7.0 and still get invalid_grant after the calls get a new refresh token.

Just saw #1357 also has the same issue with the jwt token returning previous values.

**Reply by balazsorban44** [reply]
Yeah, so turns out we were asked to increase the access_token expiry time because this problem caused the error you mentioned in production, and people were constantly asked to log in again. So actually, we did not prevent the bug, just worked around it by a longer expiry time... I'll try to think about it when I have time.

**Comment by balazsorban44** [comment]
I started to think that the current tutorial doesn't actually handle the token rotation very well... I think we might need a new endpoint that makes sure that we always return a fresh `access_token`, but I am stuck. See the PR https://github.com/nextauthjs/next-auth/pull/2062, and please come with suggestions there, if you have any good ideas! The problem is when multiple front-end requests would want to refresh the token at the same time.

**Comment by balazsorban44** [comment]
Someone found a potential issue that can lead to this problem! https://github.com/nextauthjs/next-auth/issues/2071

It is similar to what I found, there is a race condition happening. Either from the same tab or different ones.

**Comment by HarunKilic** [comment]
I've the same issue. And what i've found is that when the token gets renewed at `server side`, the client side will not get refreshed. So the `refresh_token` at the client side will remain the old state and not with the new `refresh_token` that `server side` generated. 

Opened an issue https://github.com/nextauthjs/next-auth/issues/2129 for this, as this is huge problem for those who is using the `refresh_token` approach.

**Reply by mohammedsafvan** [reply]
try disabling ssr

**Comment by EmilioHerreraSoukup** [comment]
Any update on this? Its been a long time and I cant find anything that solves it :(

**Reply by singleseeker** [reply]
Yep, it makes me crazy.

**Reply by namanwahi** [reply]
Have been facing the same issue.

**Reply by alonsourena** [reply]
Mee too

**Reply by thea-unique** [reply]
me too

**Reply by PrabeenGautam** [reply]
Did anyone solved this?

**Reply by TorrCod** [reply]
Hi did anyone fix it?, Im considering handling the token rotation on client side 
similar on this one https://www.youtube.com/watch?v=RPl0r-Yl6pU what do you guys think of this?

**Reply by Arthacus** [reply]
I’m also curious, did anyone find a solution, or got it to work in a later version? Any links or pointer in the right direction would be helpful. @avisra, good job describing the issue 🙌

**Comment by YuraZakaryan** [comment]
@avisra The problem related with twice render pages (for me), I resolve it  strictMode: false in next.config.js
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: false,
};

module.exports = nextConfig;

**Comment by emreakdas** [comment]
@HarunKilic @avisra @balazsorban44 We are experiencing this problem when using identity server 4 or 6. Have you found a solution?

**Reply by emreakdas** [reply]
@balazsorban44 help pls

**Comment by VGontier-cmd** [comment]
Any update on this ?

**Comment by FrancisVillarba** [comment]
Any update on this? This is still a recuring issue it seems :/

**Reply by mohammedsafvan** [reply]
do you tried disabling ssr

**Comment by nguyenhuugiatri** [comment]
Check my solution out:

https://github.com/nextauthjs/next-auth/discussions/3940#discussioncomment-8243238
